### PR TITLE
Extract variable resolution structure

### PR DESCRIPTION
**Pull request by alexmojaki**
Stacked on #1945.

This extracts the non-template/non-composition variable structure from the larger composition branch:

- introduce `_BaseVariable` with the existing shared variable resolution implementation
- make `Variable` a thin subclass whose `get()` delegates to `_get_result_and_record_span()`
- add `_record_exception()` for variable span exception recording and tests for the CPython traceback edge case

Intentionally excludes template variables, composition expansion, composed metadata, serialized-default composition, and template schema changes.

### CODE REVIEW

File: logfire/variables/variable.py
[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/variables/variable.py:1]: (AI) Addressed in 796e3336

File: logfire/variables/variable.py [hunk 1]
```diff
@@ -37,6 +38,15 @@
 _VARIABLE_OVERRIDES: ContextVar[dict[str, Any] | None] = ContextVar('_VARIABLE_OVERRIDES', default=None)
 
 
+def _record_exception(exception: BaseException, span: logfire.LogfireSpan) -> None:
+    """Record an exception on a span, ignoring a CPython traceback extraction bug."""
+    try:
+        span.record_exception(exception)
+    except RuntimeError as exc:
+        if 'generator raised StopIteration' not in str(exc):
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/variables/variable.py]: (AI) Addressed in 817ade88

File: logfire/variables/variable.py
Comment (alexmojaki) [inline logfire/variables/variable.py:209]: can we add and use `render_fn: Callable[[str], str] | None = None,` here for better alignment, even if nothing uses it yet

Comment (alexmojaki) [reply logfire/variables/variable.py:209]: (AI) Addressed in 611b4b16

File: logfire/variables/variable.py [hunk 2]
```diff
@@ -356,6 +321,45 @@ def _get_default(
...
+        attributes: Mapping[str, Any] | None,
+        span: logfire.LogfireSpan | None,
+        render_fn: Callable[[str], str] | None = None,
+    ) -> ResolvedVariable[T_co] | None:
+        """Resolve the code default through composition/rendering when needed."""
+        serialized_default = self._get_serialized_default(targeting_key, attributes)
+        if serialized_default is None:
+            return None
+        if render_fn is None:
+            return None
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/variables/variable.py]: (AI) Addressed in d19c2d7a

[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/variables/variable.py]: reverted

File: logfire/variables/variable.py [hunk 3]
```diff
@@ -157,11 +163,19 @@ def __init__(
...
-    def _deserialize(self, serialized_value: str) -> T_co | Exception:
+    def get_template_inputs_schema(self) -> dict[str, Any] | None:
+        """Return the JSON schema for template inputs.
+
+        Returns None on plain `Variable` instances. `TemplateVariable` overrides this
+        to return the schema derived from its `inputs_type`.
+        """
+        return None
+
+    def _deserialize(self, serialized_value: str) -> T_co | ValidationError | ValueError:
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/variables/variable.py:174]: (AI) Addressed in stacked PR #1949.

File: logfire/variables/variable.py [hunk 4]
```diff
@@ -278,34 +227,23 @@ def _resolve(
...
             if serialized_result.value is None:
+                default_result = self._resolve_serialized_default(
+                    provider,
+                    targeting_key,
+                    attributes,
+                    span,
+                    render_fn=render_fn,
+                )
+                if default_result is not None:
+                    return default_result
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/variables/variable.py:246]: (AI) Addressed in stacked PR #1949.

File: logfire/variables/variable.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/variables/variable.py]: (AI) Addressed in stacked PR #1949.

File: logfire/variables/variable.py [hunk 5]
```diff
@@ -278,34 +227,23 @@ def _resolve(
...
                 # Provider had no value; surface that the code default was used.
```
[the remark this answers is not in the document]
Comment (alexmojaki) [reply logfire/variables/variable.py:247]: (AI) Addressed in stacked PR #1949.

**Comment by dmontagu** [comment]
Superseded by #1731 — the structural extract from this branch was merged into the variable-composition branch in b2f4c768, and the architectural follow-up that David was working on (extract a shared lookup helper so composition's `resolve_ref` and `_resolve` stop drifting) landed there as 85dcbadf. Closing in favor of #1731.
